cellarium/ml/distributions/negative_binomial.py

```diff
@@ -0,0 +1,79 @@
+# Copyright Contributors to the Cellarium project.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from numbers import Number
+
+import torch
+from pyro.distributions import TorchDistribution, constraints
+from torch.distributions.utils import broadcast_all, lazy_property
+
+
+class NegativeBinomial(TorchDistribution):
+    """Negative binomial distribution.
+
+    Args:
+        mu:
+            Mean of the distribution.
+        theta:
+            Inverse dispersion.
+    """
+
+    arg_constraints = {"mu": constraints.greater_than_eq(0), "theta": constraints.greater_than_eq(0)}
+    support = constraints.nonnegative_integer
+
+    def __init__(self, mu: torch.Tensor, theta: torch.Tensor, validate_args: bool | None = None) -> None:
+        self.mu, self.theta = broadcast_all(mu, theta)
+        if isinstance(mu, Number) and isinstance(theta, Number):
+            batch_shape = torch.Size()
+        else:
+            batch_shape = self.mu.size()
+        super().__init__(batch_shape, validate_args=validate_args)
+
+    @property
+    def mean(self) -> torch.Tensor:
+        return self.mu
+
+    @property
+    def variance(self) -> torch.Tensor:
+        return (self.mu + (self.mu**2) / self.theta).masked_fill(self.theta == 0, 0)
+
+    @lazy_property
+    def _gamma(self) -> torch.distributions.Gamma:
+        # Note we avoid validating because self.theta can be zero.
+        return torch.distributions.Gamma(
+            concentration=self.theta,
+            rate=(self.theta / self.mu).masked_fill(self.theta == 0, 1.0),
+            validate_args=False,
+        )
+
+    def sample(self, sample_shape=torch.Size()) -> torch.Tensor:
+        with torch.no_grad():
+            rate = self._gamma.sample(sample_shape=sample_shape)
+            return torch.poisson(rate)
+
+    def log_prob(self, value: torch.Tensor) -> torch.Tensor:  # type: ignore[override]
+        if self._validate_args:
+            self._validate_sample(value)
+
+        # Original implementation from scVI:
+        #
+        #   log_theta_mu_eps = torch.log(self.theta + self.mu + self.eps)
+        #   return (
+        #       self.theta * (torch.log(self.theta + self.eps) - log_theta_mu_eps)
+        #       + value * (torch.log(self.mu + self.eps) - log_theta_mu_eps)
+        #       + torch.lgamma(value + self.theta)
+        #       - torch.lgamma(self.theta)
+        #       - torch.lgamma(value + 1)
+        #   )
+        delta = torch.where(
+            (value / self.theta < 1e-2) & (self.theta > 1e2),
```

Comment:
I have to admit I do not understand how this works.  Can you explain the logic here @ordabayevy ?

Reply:
@ordabayevy Actually we don't need the first condition-- my bad. Let's do this:

```
theta = torch.where(
    self.theta > THETA_THRESHOLD_STIRLING_SWITCH,
    (value + self.theta - 0.5) * torch.log1p(value / self.theta) - value,
    (value + self.theta).lgamma() - self.theta.lgamma() - torch.xlogy(value, self.theta),
)
```

and let's make `THETA_THRESHOLD_STIRLING_SWITCH = 200.` a class constant.

@sjfleming The idea is to swtich to leading order Stirling approximation for gamma when `self.theta` is large enough. If `self.theta` is large enough, then `value + self.theta` is also at least as large since `value >= 0`. The expression `(value + self.theta - 0.5) * torch.log1p(value / self.theta) - value` is indeed the leading order Stirling approximation of the exact expression `(value + self.theta).lgamma() - self.theta.lgamma() - torch.xlogy(value, self.theta)`, see [here](https://en.wikipedia.org/wiki/Stirling%27s_approximation#Stirling's_formula_for_the_gamma_function).

Reply:
Very interesting!

Reply:
@mbabadi sounds good! I have to admit that I didn't dive deep into how this approximation works but I trust you!